shared/src/ledger/ibc.rs

```diff
@@ -1,36 +1,419 @@
+                        }
+                        _ => {
+                            tracing::info!(
+                                "unexpected state change for an IBC client: \
+                                 key {}",
+                                key
+                            );
+                            false
+                        }
+                    }
+                }
+                // TODO implement validations for modules
+                IbcPrefix::Connection => false,
+                IbcPrefix::Channel => false,
+                IbcPrefix::Packet => false,
+                IbcPrefix::Unknown => {
+                    return Err(Error::KeyError(format!(
+                        "Invalid IBC-related key: {}",
+                        key
+                    )));
+                }
+            };
+            if !accepted {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+enum StateChange {
+    Created,
+    Updated,
+    Deleted,
+    NotExists,
+}
+
+enum IbcPrefix {
+    Client,
+    Connection,
+    Channel,
+    Packet,
+    Unknown,
+}
+
+impl<'a, DB, H> Ibc<'a, DB, H>
+where
+    DB: 'static + storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+{
+    /// Returns the prefix after #IBC
+    fn get_ibc_prefix(key: &Key) -> IbcPrefix {
+        match key.segments.get(1) {
+            Some(prefix) => match &*prefix.raw() {
+                "clients" => IbcPrefix::Client,
+                "connections" => IbcPrefix::Connection,
+                "channelEnds" => IbcPrefix::Channel,
+                "packets" => IbcPrefix::Packet,
+                _ => IbcPrefix::Unknown,
+            },
+            None => IbcPrefix::Unknown,
+        }
+    }
+
+    /// Returns the client ID after #IBC/clients
+    fn get_client_id(key: &Key) -> Result<ClientId> {
+        match key.segments.get(2) {
+            Some(id) => ClientId::from_str(&id.raw())
+                .map_err(|e| Error::KeyError(e.to_string())),
+            None => Err(Error::KeyError(format!(
+                "The client key doesn't have a client ID: {}",
+                key
+            ))),
+        }
+    }
+
+    fn get_client_state_change(
+        &self,
+        client_id: &ClientId,
+    ) -> Result<StateChange> {
+        let path = Path::ClientState(client_id.clone()).to_string();
+        let key = Key::ibc_key(path)
+            .expect("Creating a key for a client type failed");
+        if self.ctx.has_key_pre(&key)? {
+            if self.ctx.has_key_post(&key)? {
+                Ok(StateChange::Updated)
+            } else {
+                Ok(StateChange::Deleted)
+            }
+        } else if self.ctx.has_key_post(&key)? {
+            Ok(StateChange::Created)
+        } else {
+            Ok(StateChange::NotExists)
+        }
+    }
+
+    fn validate_created_client(&self, client_id: &ClientId) -> Result<bool> {
```

Comment:
we should actually call `ibc-rs` here to check that the client state and consensus state are semantically valid, yes?

Reply:
Do you mean that the function is [`Initialize()`](https://github.com/cosmos/ibc-go/blob/ca7f92d0a173a1a4d3fbb565cdba7aab015a5d83/modules/core/02-client/keeper/client.go#L35) in `ibc-go`?

Reply:
yes

Reply:
do we call that somewhere?

Reply:
The `Initialize()` checks if the client type is consistent between the client state and the consensus state.
`ibc-rs` wraps these states as enum [`AnyClientState`](https://github.com/informalsystems/ibc-rs/blob/47899196c271f523a1db395202d77072beaa0593/modules/src/ics02_client/client_state.rs#L44) and [`AnyConsensusState`](https://github.com/informalsystems/ibc-rs/blob/47899196c271f523a1db395202d77072beaa0593/modules/src/ics02_client/client_consensus.rs#L46), and we can check the consistency as [here](https://github.com/anomanetwork/anoma/blob/44e9ffc8f6c10961769460a301a00fc16ccdcbb2/shared/src/ledger/ibc.rs#L222).